File: apps/cli/voxcli/commands/voices.py
"""`vox voices` -- named voice profiles."""

import json
import os
import platform
import shlex
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

from voxcore import ASRClient, TTSClient


def add_parser(sub):
    p = sub.add_parser("voices", help="manage named voices")
    ops = p.add_subparsers(dest="op", required=True)

    ops.add_parser("list", help="list registered voices")

    add = ops.add_parser("add", help="register a voice from a reference recording")
    add.add_argument("id")
    source = add.add_mutually_exclusive_group(required=True)
    source.add_argument("--audio", help="reference audio file")
    source.add_argument("--record", nargs="?", const=0.0, type=float, metavar="SECONDS",
                        help="record from a microphone; optionally stop after SECONDS")
    add.add_argument("--device", help="ffmpeg audio input device (only with --record)")
    add.add_argument("--text", help="reference transcript; omitted uses the configured ASR")
    add.add_argument("--language", default="auto", help="ASR language when --text is omitted")
    add.add_argument("--edit", action="store_true",
                     help="edit the transcript with $VISUAL or $EDITOR before registering")
    add.add_argument("--dry-run", action="store_true",
                     help="print the final transcript without registering the voice")

    ops.add_parser("show", help="show one voice").add_argument("id")
    ops.add_parser("rm", help="delete a voice").add_argument("id")
    return p


def _record_command(output: Path, duration: float, device: str | None,
                    system: str | None = None) -> list[str]:
    system = system or platform.system()
    base = ["ffmpeg", "-hide_banner", "-loglevel", "error", "-y"]
    if system == "Darwin":
        source = device or "0"
        command = [*base, "-f", "avfoundation", "-i", f":{source.lstrip(':')}"]
    elif system == "Linux":
        command = [*base, "-f", "pulse", "-i", device or "default"]
    elif system == "Windows":
        command = [*base, "-f", "dshow", "-i", f"audio={device or 'default'}"]
    else:
        raise SystemExit(f"microphone recording is not supported on {system}")

    if duration:
        command.extend(["-t", str(duration)])
    command.extend(["-ac", "1", "-ar", "16000", str(output)])
    return command


def _record_audio(duration: float, device: str | None) -> Path:
    if duration < 0:
        raise SystemExit("--record duration must be greater than zero")
    if shutil.which("ffmpeg") is None:
        raise SystemExit("ffmpeg not found on PATH; install ffmpeg to use --record")

    fd, name = tempfile.mkstemp(prefix="voxstudio-voice-", suffix=".wav")
    os.close(fd)
    output = Path(name)
    command = _record_command(output, duration, device)
    try:
        if duration:
            print(f"recording for {duration:g}s...", file=sys.stderr)
            subprocess.run(command, check=True)
        else:
            process = subprocess.Popen(command, stdin=subprocess.PIPE)
            try:
                input("recording... press Enter to stop\n")
            except (EOFError, KeyboardInterrupt):
                print(file=sys.stderr)
            process.communicate(b"q\n")
            if process.returncode:
                raise subprocess.CalledProcessError(process.returncode, command)
        if output.stat().st_size <= 44:
            raise SystemExit("recording produced no audio")
        print(f"recorded {output}", file=sys.stderr)
        return output
    except (OSError, subprocess.CalledProcessError) as exc:
        raise SystemExit(f"recording failed: {exc}") from exc
    finally:
        if output.exists() and output.stat().st_size <= 44:
            output.unlink()


def _edit_transcript(text: str) -> str:
    editor = os.environ.get("VISUAL") or os.environ.get("EDITOR")
    if not editor:
        raise SystemExit("--edit requires $VISUAL or $EDITOR")

    path = None
    try:
        with tempfile.NamedTemporaryFile("w", encoding="utf-8", suffix=".txt", delete=False) as fh:
            fh.write(text)
            if text and not text.endswith("\n"):
                fh.write("\n")
            path = Path(fh.name)
        subprocess.run([*shlex.split(editor), str(path)], check=True)
        return path.read_text(encoding="utf-8").strip()
    except (OSError, subprocess.CalledProcessError) as exc:
        raise SystemExit(f"editor failed: {exc}") from exc
    finally:
        if path is not None:
            path.unlink(missing_ok=True)


def _voice_transcript(args, cfg, audio: str) -> str:
    if args.text is not None:
        text = args.text.strip()
    else:
        with ASRClient(cfg.engine("asr")) as asr:
            result = asr.transcribe(audio, language=args.language)
        text = result.text.strip()
        language = result.lang or "unknown"
        print(f"ASR transcript ({language}): {text}", file=sys.stderr)

    if args.edit:
        text = _edit_transcript(text)
    if not text:
        raise SystemExit("reference transcript is empty")
    return text


def run(args, cfg) -> int:
    if args.op == "add":
        if args.device and args.record is None:
            raise SystemExit("--device requires --record")
        recording = _record_audio(args.record, args.device) if args.record is not None else None
        audio = str(recording) if recording else args.audio
        completed = False
        try:
            text = _voice_transcript(args, cfg, audio)
            if args.dry_run:
                print(text)
                completed = True
                return 0

            with TTSClient(cfg.engine("tts"), cfg.tts_defaults) as tts:
                result = tts.create_voice(args.id, text, audio)
            print(json.dumps(result, ensure_ascii=False))
            completed = True
            return 0
        finally:
            if recording:
                if completed:
                    recording.unlink(missing_ok=True)
                else:
                    print(f"recording kept at {recording}", file=sys.stderr)

    with TTSClient(cfg.engine("tts"), cfg.tts_defaults) as tts:
        if args.op == "list":
            voices = tts.list_voices()
            if not voices:
                print("(no registered voices)")
            for v in voices:
                print(f"{v['id']:<20} {v.get('prompt_audio_length', '?')}s  {v.get('created_at', '')}")
        elif args.op == "show":
            print(json.dumps(tts.get_voice(args.id), ensure_ascii=False))
        elif args.op == "rm":
            tts.delete_voice(args.id)
            print(f"deleted {args.id}")
    return 0

File: apps/cli/src/commands/voices.ts
import { basename } from "node:path";
import { AsrClient, TtsClient, type Fetch } from "@voxstudio/clients";
import { engine } from "@voxstudio/config";
import type { VoxConfig } from "@voxstudio/contracts";
import {
  editText,
  readFileBlob,
  recordAudio,
  removeRecording,
} from "@voxstudio/platform-bun";
import type { CliIo } from "../io";

interface AddOptions {
  operation: "add";
  id: string;
  audio?: string;
  record?: number;
  device?: string;
  text?: string;
  language: string;
  edit: boolean;
  dryRun: boolean;
}

type VoiceOptions =
  | AddOptions
  | { operation: "list" }
  | { operation: "show" | "rm"; id: string };

export interface VoicePlatform {
  editText(text: string): Promise<string>;
  recordAudio(duration: number, device: string | undefined, status: (message: string) => void): Promise<string>;
  removeRecording(path: string): Promise<void>;
}

const defaultPlatform: VoicePlatform = { editText, recordAudio, removeRecording };

export const voicesUsage = `usage: vox voices {list,add,show,rm} ...

Manage named voices.

commands:
  list                         list registered voices
  add ID --audio PATH          register from an audio file
  add ID --record [SECONDS]    register from microphone input
  show ID                      show one voice
  rm ID                        delete one voice

add options:
  --device NAME       ffmpeg input device (only with --record)
  --text TEXT         reference transcript; omitted uses ASR
  --language LANG     ASR language (default: auto)
  --edit              edit transcript with $VISUAL or $EDITOR
  --dry-run           print transcript without registering`;

function value(args: string[], index: number, option: string): string {
  const result = args[index];
  if (!result) throw new TypeError(`voices: ${option} requires a value`);
  return result;
}

function parseAdd(args: string[]): AddOptions {
  let id: string | undefined;
  let audio: string | undefined;
  let record: number | undefined;
  let device: string | undefined;
  let text: string | undefined;
  let language = "auto";
  let edit = false;
  let dryRun = false;
  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index] as string;
    if (arg === "--audio") audio = value(args, ++index, arg);
    else if (arg === "--record") {
      const candidate = args[index + 1];
      if (candidate !== undefined && (!candidate.startsWith("-") || Number.isFinite(Number(candidate)))) {
        record = Number(candidate);
        if (!Number.isFinite(record)) throw new TypeError("voices: --record must be a number");
        index += 1;
      } else record = 0;
    } else if (arg === "--device") device = value(args, ++index, arg);
    else if (arg === "--text") text = value(args, ++index, arg);
    else if (arg === "--language") language = value(args, ++index, arg);
    else if (arg === "--edit") edit = true;
    else if (arg === "--dry-run") dryRun = true;
    else if (arg.startsWith("-")) throw new TypeError(`voices: unknown option ${arg}`);
    else if (id === undefined) id = arg;
    else throw new TypeError("voices add: expected one voice ID");
  }
  if (!id) throw new TypeError("voices add: voice ID is required");
  if ((audio === undefined) === (record === undefined)) {
    throw new TypeError("voices add: exactly one of --audio or --record is required");
  }
  if (device !== undefined && record === undefined) throw new TypeError("--device requires --record");
  if (record !== undefined && record < 0) throw new TypeError("--record duration must be non-negative");
  return {
    operation: "add", id, language, edit, dryRun,
    ...(audio === undefined ? {} : { audio }),
    ...(record === undefined ? {} : { record }),
    ...(device === undefined ? {} : { device }),
    ...(text === undefined ? {} : { text }),
  };
}

function parse(args: string[]): VoiceOptions {
  const operation = args.shift();
  if (operation === "list") {
    if (args.length) throw new TypeError("voices list: no arguments expected");
    return { operation };
  }
  if (operation === "add") return parseAdd(args);
  if (operation === "show" || operation === "rm") {
    if (args.length !== 1) throw new TypeError(`voices ${operation}: one voice ID is required`);
    return { operation, id: args[0] as string };
  }
  throw new TypeError("voices: expected list, add, show, or rm");
}

export async function runVoices(
  args: string[],
  config: VoxConfig,
  io: CliIo,
  fetch: Fetch = globalThis.fetch,
  platform: VoicePlatform = defaultPlatform,
): Promise<number> {
  const options = parse([...args]);
  if (options.operation === "add") {
    let recording: string | undefined;
    let completed = false;
    try {
      recording = options.record === undefined
        ? undefined
        : await platform.recordAudio(options.record, options.device, io.err);
      const audioPath = recording ?? options.audio as string;
      let audio: Blob | undefined;
      let transcript: string;
      if (options.text !== undefined) transcript = options.text.trim();
      else {
        audio = await readFileBlob(audioPath);
        const result = await new AsrClient(engine(config, "asr"), fetch)
          .transcribe(audio, basename(audioPath), options.language);
        transcript = result.text.trim();
        io.err(`ASR transcript (${result.lang ?? "unknown"}): ${transcript}`);
      }
      if (options.edit) transcript = await platform.editText(transcript);
      if (!transcript) throw new TypeError("reference transcript is empty");
      if (options.dryRun) io.out(transcript);
      else {
        audio ??= await readFileBlob(audioPath);
        const voice = await new TtsClient(engine(config, "tts"), fetch)
          .createVoice(options.id, transcript, audio, basename(audioPath));
        io.out(JSON.stringify(voice));
      }
      completed = true;
      return 0;
    } finally {
      if (recording) {
        if (completed) await platform.removeRecording(recording);
        else io.err(`recording kept at ${recording}`);
      }
    }
  }

  const tts = new TtsClient(engine(config, "tts"), fetch);
  if (options.operation === "list") {
    const voices = await tts.listVoices();
    if (!voices.length) io.out("(no registered voices)");
    for (const voice of voices) {
      io.out(`${voice.id.padEnd(20)} ${voice.prompt_audio_length ?? "?"}s  ${voice.created_at ?? ""}`);
    }
  } else if (options.operation === "show") {
    io.out(JSON.stringify(await tts.getVoice(options.id)));
  } else {
    await tts.deleteVoice(options.id);
    io.out(`deleted ${options.id}`);
  }
  return 0;
}
